libcudacxx/include/cuda/std/__cccl/assert.h

```diff
@@ -0,0 +1,123 @@
+#    else // ^^^ _CCCL_CUDA_COMPILER_NVCC ^^^ / vvv !_CCCL_CUDA_COMPILER_NVCC vvv
+#      if __has_include(<__assert>) // libc++ uses _LIBCPP_ASSERT from <__assert>
+#        include <__assert>
+#        define _CCCL_ASSERT_IMPL(expression, message) _LIBCPP_ASSERT(expression, message)
+#      else // libstdc++ uses __glibcxx_assert from <bits/c++config.h
+#        include <bits/c++config.h>
+#        if defined(_GLIBCXX_RELEASE) && _GLIBCXX_RELEASE >= 12
+// libstdc++ does not fully qualify its use of `__is_constant_evaluated`
+// It was introduced in the assert handling in 5e8a30d
+// libstdc++ : Redefine __glibcxx_assert to work in C++ 23 constexpr
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+using ::std::__is_constant_evaluated;
+_LIBCUDACXX_END_NAMESPACE_STD
+#        endif // _GLIBCXX_RELEASE >= 12
+#        define _CCCL_ASSERT_IMPL(expression, message) __glibcxx_assert(expression)
+#      endif // libstdc++
+#    endif // !_CCCL_CUDA_COMPILER_NVCC
+#  else // ^^^ CCCL_ENABLE_DEVICE_ASSERTIONS ^^^ / vvv !CCCL_ENABLE_DEVICE_ASSERTIONS vvv
+#    define _CCCL_ASSERT_IMPL
+#  endif // !CCCL_ENABLE_DEVICE_ASSERTIONS
+#endif // __CUDA_ARCH__
+
+//! Use the different standard library implementations to implement host side asserts
+#ifndef __CUDA_ARCH__
+#  ifdef CCCL_ENABLE_HOST_ASSERTIONS
+#    if __has_include(<yvals_core.h>) // MSVC uses _STL_VERIFY from <yvals_core.h>
+#      include <yvals_core.h>
+#      define _CCCL_ASSERT_IMPL(expression, message) _STL_VERIFY(expression, message)
+#    elif __has_include(<__assert>) // libc++ uses _LIBCPP_ASSERT from <__assert>
+#      include <__assert>
+#      define _CCCL_ASSERT_IMPL(expression, message) _LIBCPP_ASSERT(expression, message)
+#    else // libstdc++ uses __glibcxx_assert from <bits/c++config.h
+#      include <bits/c++config.h>
+#      if defined(_GLIBCXX_RELEASE) && _GLIBCXX_RELEASE >= 12
+// libstdc++ does not fully qualify its use of `__is_constant_evaluated`
+// It was introduced in the assert handling in 5e8a30d
+// libstdc++ : Redefine __glibcxx_assert to work in C++ 23 constexpr
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+using ::std::__is_constant_evaluated;
+_LIBCUDACXX_END_NAMESPACE_STD
+#      endif // _GLIBCXX_RELEASE >= 12
+#      define _CCCL_ASSERT_IMPL(expression, message) __glibcxx_assert(expression)
+#    endif // libstdc++
+#  endif // !CCCL_ENABLE_HOST_ASSERTIONS
+#endif // !__CUDA_ARCH__
+
+//! _CCCL_VERIFY is an unconditionally enabled assertion that is reserved for the most important checks
+//! Some compilers warn about `is_constant_evaluated()` in a non constexpr function, so silence that right away
+#if defined(_CCCL_COMPILER_ICC)
+#  define _CCCL_VERIFY(expression, message) \
+    _CCCL_DIAG_PUSH _CCCL_DIAG_SUPPRESS_ICC(4190) _CCCL_ASSERT_IMPL(expression, message) _CCCL_DIAG_POP
+#else // ^^^ _CCCL_COMPILER_ICC ^^^ / vvv !_CCCL_COMPILER_ICC vvv
+#  define _CCCL_VERIFY(expression, message) _CCCL_ASSERT_IMPL(expression, message)
```

Comment:
If `__CUDA_ARCH__` is defined but `CCCL_ENABLE_DEVICE_ASSERTIONS` is not defined, then `_CCCL_ASSERT_IMPL` will be defined as empty, so `_CCCL_VERIFY(expression, message)` will expand to `(expression, message)`.  That will compile and behave correctly, but it's not what you want.

If neither `__CUDA_ARCH__` nor `CCCL_ENABLE_HOST_ASSERTIONS` are defined, then `_CCCL_ASSERT_IMPL` is not defined at all, so `_CCCL_VERIFY(expression, message)` will expand to `_CCCL_ASSERT_IMPL(expression, message)`, which is a compilation error.

Since `_CCCL_VERIFY` is always defined, `_CCCL_ASSERT_IMPL` must always be defined.  If it can't be defined as something meaningful, then it should be `#define _CCCL_ASSERT_IMPL(expression, message) ((void)0)`